include/swift/Basic/Located.h

```diff
@@ -0,0 +1,39 @@
+//===--- Located.h - Source Location and Associated Value ----------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file forward declares and imports various common LLVM datatypes that
+// swift wants to use unqualified.
+//
+//===----------------------------------------------------------------------===//
+
+
+#ifndef SWIFT_BASIC_LOCATED_H
+#define SWIFT_BASIC_LOCATED_H
+#include "swift/Basic/SourceLoc.h"
+
+namespace swift {
+
+template<typename T>
+struct Located {
+
+  T item;
+
+  SourceLoc loc;
+
+  template<typename U>
+  friend bool operator ==(const Located<U> lhs, const Located<U> rhs) {
+    return lhs.item == rhs.item && lhs.loc == rhs.loc;
+  }
+};
```

Comment:
Could you add a `dump` method that does something reasonable like for other types? You could have two overloads: `void dump() const; void dump(llvm::raw_ostream &OS) const` where the former calls the latter using `llvm::errs()`. [For example, `SourceFile` does this.]

You could add a new file `Located.cpp` which has the implementations and the appropriate `#includes`. You will also need to add `Located.cpp` to the `lib/Basic/CMakeLists.txt` file.

Reply:
@kitaisreal what do you think about this point? I think it might be useful to have this method for debugging.

As to what is should print, perhaps something that prints the SourceLoc first and then the item, leaving a space in between (say)

Reply:
@varungandhi-apple I agree, that we should introduce such function. Added.